Buy album

shernanm opened the pull request:
## Summary

* The endpoint that allows you to buy an album is implemented.
* Their respective tests are carried out.
* Mocks are made for the response of the external service.

## Trello Card

https://trello.com/c/aUMjjbMh

Commit:
API consumption of albums is added

Commit:
Added code for a bad request

Commit:
Adding user model

Commit:
Changes requested to API consumption

Commit:
Merge branch 'api-integration' of https://github.com/wolox-training/shm-express-js into user-model

Commit:
wip

Commit:
merge with master

Commit:
Adjustments were made to the model and migration

Commit:
wip

Commit:
Changes are made in the definition of the model and migration.

Commit:
The deleted_at property is added to the migration and configurations are made in the model.

Commit:
Merge branch 'user-model' into user-sign-up

Commit:
wip

Commit:
The deleted_at property is modified

Commit:
Merge branch 'user-model' into user-sign-up

Commit:
The user registry was implemented.

Commit:
Added tests and documentation for the endpoint of users.

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
The use of middleware is implemented for the endpoints of albums and users.

Commit:
wip

Commit:
Modification is made to only have a schema validator.

Commit:
wip

Commit:
Merge branch 'master' into middlewares

Commit:
Changes are made to the tests with the use of middlewares

Commit:
Correcting the salt rounds of the encryption and the message sent in the user creation

Commit:
wip

Commit:
Merge branch 'middlewares' into user-sign-in

Commit:
User login is implemented and its respective test.

Commit:
The endpoint is implemented that allows to list all the users who registered

Commit:
Added test for user list.

Commit:
wip

Commit:
The used library was changed to generate a token.

Commit:
merge with user-sign-in

Commit:
The way to validate the tokens is modified.

Commit:
merge with get-user-list.

Commit:
The creation of administrator users is implemented.

Commit:
wip

Commit:
The signInUser method is modified.

Commit:
merge with user-sign-in

Commit:
merge with get-user-list

Commit:
The request library is added.

Commit:
Added sequelize CLI.

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
The endpoint is implemented that allows a user to buy an album.

Commit:
Merge branch 'master' into buy-album

Commit:
Improvements are made to the endpoint that allows you to buy albums and tests in general

Commit:
Middleware is added to the endpoint that allows you to buy albums

raparicio6 commented on app/controllers/album.js:
```diff
@@ -1,11 +1,13 @@
+const errors = require('../errors');
 
-exports.getAlbums = (req, res, next) => {
+exports.getAllAlbums = (req, res, next) => {
```
its ok to call getAlbums in order to make the endpoint more generic. you may want to get all the albums sometimes but other times you may only want some albums passing a filter.
it wont be good to create two different endpoints for doing that

raparicio6 commented on app/controllers/album.js:
```diff
@@ -18,12 +20,37 @@ exports.getPhotos = (req, res, next) => {
 };
+
+exports.buyAlbums = (req, res, next) => {
+  const qs = {
```
why do you do this?
you can pass req.params to findAlbumBy

shernanm replied:
I did it so the method is general, plus I'm also using it below in getAlbums.

raparicio6 replied:
Its ok but not neccesary. You are just copying req.params object into a new one called qs

raparicio6 commented on app/controllers/album.js:
```diff
@@ -18,12 +20,37 @@ exports.getPhotos = (req, res, next) => {
+    id: req.params.id
+  };
+  logger.info(`buyAlbums method start, request methods: ${req.method}, endpoint: ${req.path}, id: ${qs.id}`);
+  findAlbumBy(qs)
```
add return

raparicio6 commented on app/controllers/album.js:
```diff
@@ -18,12 +20,37 @@ exports.getPhotos = (req, res, next) => {
+        return next(errors.buyAlbumError('Duplicate purchase of an album is not allowed'));
+      }
+      const user = decodedToken(req.headers.token);
+      return getAlbums(qs).then(([{ id, title }]) => {
```
remove { }

raparicio6 commented on app/controllers/album.js:
```diff
@@ -18,12 +20,37 @@ exports.getPhotos = (req, res, next) => {
+      return getAlbums(qs).then(([{ id, title }]) => {
+        albumRegister(albumMapper(id, title, user.id)).then(() => {
+          logger.info(`Album ${title} successfully purchased`);
+          res.status(201).send({
```
add return

raparicio6 commented on app/services/users.js:
```diff
@@ -69,3 +69,43 @@ exports.changeRole = (role, email) =>
+  });
+
+exports.AdminUserRegister = user =>
+  User.create({ ...user, role: 'admin' }).catch(() => {
```
why another method for admin user?
if you have 10 roles you would create 10 methods?
try to reuse the original create user

shernanm replied:
It was my mistake, I was using it but in the end reuse userRegister.

Commit:
Modifications are made to the endpoint to buy an album and for the tests

raparicio6 commented on app/mappers/mappers.js:
```diff
@@ -0,0 +1,16 @@
+exports.mapperUserList = ({ count, rows }, limit, page) => {
```
dont create a big `mapper` file. 
create one file per controller

raparicio6 commented on package.json:
```diff
@@ -64,9 +64,12 @@
     "jsonwebtoken-promisified": "^1.0.3",
     "jwt-simple": "^0.5.1",
     "morgan": "^1.9.1",
+    "nock": "^10.0.6",
```
you need this?

raparicio6 commented on test/album.test.js:
```diff
@@ -0,0 +1,65 @@
+const request = require('supertest');
+const dictum = require('dictum.js');
+const { albumMock, errorAlbumMock } = require('./mocks/albums');
+
```
why the blank lines?

raparicio6 commented on test/mocks/albums.js:
```diff
@@ -0,0 +1,15 @@
+
+exports.albumMock = (value = albumResponse) => requestPromise.mockResolvedValue(value);
+exports.errorAlbumMock = () =>
+  requestPromise.mockRejectedValue(new Error(errors.albumError(MESSAGE_ALBUM_API_FAILED)));
```
do you need the Error object? We never use it

Commit:
Changes required

Mapper files are created per controller.

Blank lines, error object and unused library are deleted.
